components/file_system/src/iosnoop/biosnoop.rs

```diff
@@ -146,52 +145,23 @@ pub fn get_io_type() -> IOType {
     unsafe { *IDX.with(|idx| IO_TYPE_ARRAY[idx.0]) }
 }
 
-unsafe fn get_io_stats() -> Option<HashMap<IOType, IOStats>> {
-    if let Some(ctx) = BPF_CONTEXT.as_mut() {
-        let mut map = HashMap::default();
-        for e in ctx.stats_table.iter() {
-            let io_type = ptr::read(e.key.as_ptr() as *const IOType);
-            let read = std::intrinsics::atomic_load(e.value.as_ptr() as *const u64);
-            let write = std::intrinsics::atomic_load((e.value.as_ptr() as *const u64).add(1));
-            map.insert(io_type, IOStats { read, write });
-        }
-        Some(map)
-    } else {
-        None
-    }
-}
-
-pub struct IOContext {
-    io_stats_map: Option<HashMap<IOType, IOStats>>,
-}
-
-impl IOContext {
-    pub fn new() -> Self {
-        IOContext {
-            io_stats_map: unsafe { get_io_stats() },
-        }
-    }
-
-    pub fn delta_and_refresh(&mut self) -> HashMap<IOType, IOStats> {
-        if self.io_stats_map.is_some() {
-            if let Some(map) = unsafe { get_io_stats() } {
-                for (io_type, stats) in &map {
-                    self.io_stats_map
-                        .as_mut()
-                        .unwrap()
-                        .entry(*io_type)
-                        .and_modify(|e| {
-                            e.read = stats.read - e.read;
-                            e.write = stats.write - e.write;
-                        })
-                        .or_insert(stats.clone());
-                }
-
-                return self.io_stats_map.replace(map).unwrap();
+pub(crate) fn fetch_io_bytes(mut io_type: IOType) -> IOBytes {
+    unsafe {
+        if let Some(ctx) = BPF_CONTEXT.as_mut() {
+            let io_type_buf_ptr = &mut io_type as *mut IOType as *mut u8;
+            let mut io_type_buf =
+                std::slice::from_raw_parts_mut(io_type_buf_ptr, std::mem::size_of::<IOType>());
+            if let Ok(e) = ctx.stats_table.get(&mut io_type_buf) {
+                // TODO: the returned slice could be misaligned
```

Comment:
If this cannot figure the misalignment, why change that instead of just removing the atomic load.

Reply:
After reading [this](https://dwrensha.github.io/capnproto-rust/2020/01/11/unaligned-memory-access.html) and [this](https://stackoverflow.com/questions/42499049/transmuting-u8-buffer-to-struct-in-rust), it seems that `ptr::read` is vulnerable to misalignment, `transmute` or `T::from_ne_bytes` (which essentially uses transmute) is safe because they copy the value, `align_to` can be used to determine the origin slice's alignment. So @brson 's code is solid, we can remove the comment.

Reply:
Agree. This resolve the misalignment. But I prefer `ptr::read_unaligned(e.value.as_ptr() as *const IOStats)` which is neat.

Reply:
Yah it's better, making the choice of method more explicit.